plugin/pkg/up/up.go

```diff
@@ -3,80 +3,49 @@
 package up
 
 import (
-	"sync"
+	"sync/atomic"
 	"time"
 )
 
 // Probe is used to run a single Func until it returns true (indicating a target is healthy). If an Func
 // is already in progress no new one will be added, i.e. there is always a maximum of 1 checks in flight.
 type Probe struct {
-	do   chan Func
-	stop chan bool
-
-	target string
-
-	sync.Mutex
-	inprogress int
+	inprogress int32
+	interval   time.Duration
 }
 
 // Func is used to determine if a target is alive. If so this function must return nil.
 type Func func() error
 
 // New returns a pointer to an intialized Probe.
-func New() *Probe {
-	return &Probe{stop: make(chan bool), do: make(chan Func)}
-}
+func New() *Probe { return &Probe{} }
 
 // Do will probe target, if a probe is already in progress this is a noop.
-func (p *Probe) Do(f Func) { p.do <- f }
-
-// Stop stops the probing.
-func (p *Probe) Stop() { p.stop <- true }
-
-// Start will start the probe manager, after which probes can be initialized with Do.
-func (p *Probe) Start(interval time.Duration) { go p.start(interval) }
-
-func (p *Probe) start(interval time.Duration) {
-	for {
-		select {
-		case <-p.stop:
-			p.Lock()
-			p.inprogress = stop
-			p.Unlock()
-			return
-		case f := <-p.do:
-			p.Lock()
-			if p.inprogress == active || p.inprogress == stop {
-				p.Unlock()
-				continue
-			}
-			p.inprogress = active
-			p.Unlock()
-
-			// Passed the lock. Now run f for as long it returns false. If a true is returned
-			// we return from the goroutine and we can accept another Func to run.
-			go func() {
-				for {
-					if err := f(); err == nil {
-						break
-					}
-					time.Sleep(interval)
-					p.Lock()
-					if p.inprogress == stop {
-						p.Unlock()
-						return
-					}
-					p.Unlock()
+func (p *Probe) Do(f Func) {
+	if atomic.CompareAndSwapInt32(&p.inprogress, idle, active) {
```

Comment:
Coming back to this, I'm not a fan of this sync/atomic usage; it less clear than the mutexes, probably not faster (if that would be a consideration) and requires 64 bit alignment of the fields.

If we call Do() after Stop() it would call that a mistake of the caller; but bluntly hanging is not use.
Can't we just check `p.inprogress`? When starting or not call Do after Stop from the higher layers?

Reply:
controlling `Do()` and `Stop()` at higher level is not so easy, since these calls are initiated from ServeDNS() and OnShutdown() which can be called by caddy in **any** order. 
I thing controlling `Probe.Do()` is much easier. I'm going to submit another PR with minimal changes to Probe, which will check `p.inprogress` in `Do()`

Reply:
I thought about change to original code like
```
func (p *Probe) Do(f Func) {
  p.Lock()
  if p.inprogress == idle {
    p.do <- f
  }
  p.Unlock()
}
```
but it (unexpectedly) leads to deadlock

I can replace atomics with mutex in this PR if you wish. BTW, 64-bit alignment is required only for 64-bit values (see Bugs at https://golang.org/pkg/sync/atomic/), so int32 should work fine

Or you can just close this PR if you think the original code is good enough